htsohm/htsohm.py

```diff
@@ -455,24 +456,42 @@ def worker_run_loop(run_id):
                 session.add(material)
 
             if config['mutation_strength_method'] != 'flat':
-                if (
-                    material.generation_index == config['children_per_generation'] - 1
-                    and gen > 0
-                    ):
-                    parent_ids = get_all_parent_ids(run_id, gen)
-    
-                    print_block('CALCULATING MUTATION STRENGTHS')
-                    ms_bins = []
-                    for parent_id in parent_ids:
-                        parent_bin = session.query(Material).get(parent_id).bin
-                        if parent_bin not in ms_bins:
-                            print(
-                                    (
-                                        'Calculating bin-mutation-strength for bin : {0}'
-                                    ).format(parent_bin)
-                                )
-                            calculate_mutation_strength(run_id, gen + 1, parent_bin)
-                        ms_bins.append(parent_bin)
+                if material.generation_index == config['children_per_generation'] - 1 and gen > 0:
+                # standard calculation of mutation strengths for all accessed bins
+                    if gen % config['annealing_frequency'] != 0:
+                        parent_ids = get_all_parent_ids(run_id, gen)
+                        print_block('CALCULATING MUTATION STRENGTHS')
+                        ms_bins = []
+                        for parent_id in parent_ids:
+                            parent_bin = session.query(Material).get(parent_id).bin
+                            if parent_bin not in ms_bins:
+                                print('Calculating bin-mutation-strength for bin : {}' \
+                                        .format(parent_bin))
+                                calculate_mutation_strength(run_id, gen + 1, parent_bin)
+                            ms_bins.append(parent_material.bin)
+                    # annealing to reset all mutation strengths to initial value
+                    else:
+                        all_accessed_bin_strings = session \
+                                .query(func.distinct(
+                                    Material.gas_adsorption_bin,
+                                    Material.surface_area_bin,
+                                    Material.void_fraction_bin)) \
+                                .filter(
+                                    Material.run_id == run_id,
+                                    Material.retest_passed != False,
+                                    Material.generation_index < config['children_per_generation']) \
+                                .all()
+                        all_accessed_bin_tuples = [ast.literal_eval(e[0])
+                                for e in all_accessed_bin_strings]
```

Comment:
I'm sure there is a way to do this without using an eval. Have you tried not using func.distinct, and instead replacing all() with distinct() (or maybe it is distinct().all())?

btw, the rough rule of thumb is: never, ever, ever use eval except when you need to except that you never need to because it is basically always the worst possible way to do something... ;)